src/graphml.rs

```diff
@@ -0,0 +1,667 @@
+                },
+                Event::End(ref e) => match e.name() {
+                    b"key" => {
+                        matches!(state, State::Key);
+                        state = State::Start;
+                    }
+                    b"default" => {
+                        matches!(state, State::DefaultForKey);
+                        state = State::Key;
+                    }
+                    b"graph" => {
+                        matches!(state, State::Graph);
+                        state = State::Start;
+                    }
+                    b"node" => {
+                        matches!(state, State::Node);
+                        state = State::Graph;
+                    }
+                    b"edge" => {
+                        matches!(state, State::Edge);
+                        state = State::Graph;
+                    }
+                    b"data" => {
+                        matches!(state, State::DataForNode | State::DataForEdge);
+                        match state {
+                            State::DataForNode => state = State::Node,
+                            State::DataForEdge => state = State::Edge,
+                            _ => {
+                                // in all other cases we have already bailed out in `matches`.
+                                unreachable!()
+                            }
+                        }
+                    }
+                    _ => {}
+                },
+                Event::Text(ref e) => match state {
+                    State::DefaultForKey => {
+                        graphml.last_key_set_value(
+                            e.unescape_and_decode(&reader)?,
+                            domain_of_last_key,
+                        )?;
+                    }
+                    State::DataForNode => {
+                        graphml
+                            .last_node_set_data(&last_data_key, e.unescape_and_decode(&reader)?)?;
+                    }
+                    State::DataForEdge => {
+                        graphml
+                            .last_edge_set_data(&last_data_key, e.unescape_and_decode(&reader)?)?;
+                    }
+                    _ => {}
+                },
+                Event::Eof => {
+                    break;
+                }
+                _ => {}
+            }
+
+            buf.clear();
+        }
+
+        Ok(graphml)
+    }
+}
+
+/// Read a list of graphs from a file in GraphML format.
+///
+/// GraphML is a comprehensive and easy-to-use file format for graphs. It consists
+/// of a language core to describe the structural properties of a graph and a flexible
+/// extension mechanism to add application-specific data.
+///
+/// For more information see:
+/// http://graphml.graphdrawing.org/
+///
+/// .. note::
+///
+///     This implementation does not support mixed graphs (directed and unidirected edges together),
+///     hyperedges, nested graphs, or ports.
```

Comment:
I'm wondering about nested graphs. I think it should be possible to do this in retworkx as we can make the node payload a graph object. But I think we can also just save that as a future todo because just having a single layer of graphs is more than enough.

Reply:
Not sure this will work since in the following example (which is valid GraphML syntax)
```xml
<graph id="G" edgedefault="undirected">
    <node id="n0">
        <graph id="n0:" edgedefault="undirected">
          <node id="n0::n0"/>
          <node id="n0::n1"/>
          <edge id="en" source="n0::n0" target="n0::n1"/>
        </graph>
    </node>
    <node id="n1"/>
    <edge id="e" source="n0::n1" target="n1"/>
</graph>
```
we have no way to express the last edge (`<edge id="e" source="n0::n1" target="n1"/>` ) since node `"n0::n1"` will belong to the graph inside the node data payload.

We might be able to support nested graphs by returning a graph for each nesting hierarchy, e.g in the above example we  return two graphs:
1. `G0 = (V0, E0)`, with `V0 = {"n0", "n1"}` and `E0 = { ("n0", "n1") }`
2. `G1 = (V1, E1)`, with `V1 = {"n0:n0", "n0:n1", "n1"}` and `E1 = { ("n0:n0", "n1") }`

but yeah, let's see if there is really the need for nested graphs and think more thoroughly about it in the future

Reply:
Ah ok, I didn't realize that graphml let you have edges between levels in the hierarchy of nested graphs. Then yeah it's definitely not as trivial as I was thinking/hoping.